sif/internal/sif_util.go

```diff
@@ -0,0 +1,238 @@
+	if err != nil {
+		return SifImage{}, errors.Wrap(err, "looking up rootfs from SIF file")
+	}
+
+	// look for a definition file object
+	searchDesc := sif.Descriptor{Datatype: sif.DataDeffile}
+	resultDescs, _, err := image.fimg.GetFromDescr(searchDesc)
+	if err == nil && resultDescs != nil {
+		// we assume in practice that typical SIF files don't hold multiple deffiles
+		image.deffile = resultDescs[0]
+		image.defReader = io.NewSectionReader(image.fimg.Fp, image.deffile.Fileoff, image.deffile.Filelen)
+	}
+	if err = image.generateConfig(); err != nil {
+		return SifImage{}, err
+	}
+
+	// look for an environment variable set object
+	searchDesc = sif.Descriptor{Datatype: sif.DataEnvVar}
+	resultDescs, _, err = image.fimg.GetFromDescr(searchDesc)
+	if err == nil && resultDescs != nil {
+		// we assume in practice that typical SIF files don't hold multiple EnvVar sets
+		image.env = resultDescs[0]
+		image.envReader = io.NewSectionReader(image.fimg.Fp, image.env.Fileoff, image.env.Filelen)
+	}
+
+	return image, nil
+}
+
+func (image *SifImage) parseEnvironment(scanner *bufio.Scanner) error {
+	for scanner.Scan() {
+		s := strings.TrimSpace(scanner.Text())
+		if s == "" || strings.HasPrefix(s, "#") {
+			continue
+		}
+		if strings.HasPrefix(s, "%") {
+			return nil
+		}
+		image.envlist = append(image.envlist, s)
+	}
+	if err := scanner.Err(); err != nil {
+		return errors.Wrap(err, "parsing environment from SIF definition file object")
+	}
+	return nil
+}
+
+func (image *SifImage) parseRunscript(scanner *bufio.Scanner) error {
+	for scanner.Scan() {
+		s := strings.TrimSpace(scanner.Text())
+		if strings.HasPrefix(s, "%") {
+			return nil
+		}
+		image.cmdlist = append(image.cmdlist, s)
+	}
+	if err := scanner.Err(); err != nil {
+		return errors.Wrap(err, "parsing runscript from SIF definition file object")
+	}
+	return nil
+}
+
+func (image *SifImage) generateRunscript() error {
+	base := `#!/bin/bash
+`
+	image.runscript = bytes.NewBufferString(base)
+	for _, s := range image.envlist {
+		_, err := image.runscript.WriteString(fmt.Sprintln(s))
+		if err != nil {
+			return errors.Wrap(err, "writing to runscript buffer")
+		}
+	}
+	for _, s := range image.cmdlist {
+		_, err := image.runscript.WriteString(fmt.Sprintln(s))
+		if err != nil {
+			return errors.Wrap(err, "writing to runscript buffer")
+		}
+	}
+	return nil
+}
+
+func (image *SifImage) generateConfig() error {
+	if image.deffile == nil {
+		image.cmdlist = append(image.cmdlist, "bash")
+		return nil
+	}
+
+	// extract %environment/%runscript from definition file
+	var err error
+	scanner := bufio.NewScanner(image.defReader)
+	for scanner.Scan() {
+		s := strings.TrimSpace(scanner.Text())
+	again:
+		if s == `%environment` {
+			if err = image.parseEnvironment(scanner); err != nil {
+				return err
+			}
+		} else if s == `%runscript` {
+			if err = image.parseRunscript(scanner); err != nil {
+				return err
+			}
+		}
+		s = strings.TrimSpace(scanner.Text())
```

Comment:
This seems to be redundant with the line at the top of the loop at a first glance, and is very non-obvious. At the very least
```go
for scanner.Scan() {
again:
	// parseEnvironment and parseRunscript, will, on success, scan one line past its section, so we need to
	// return here without going through scanner.Scan()
	s := strings.TrimSpace(scanner.Text())
	if s == `%environment` {
		if err = image.parseEnvironment(scanner); err != nil {
			return err
		}
		goto again
	} else if s == `%runscript` {
		if err = image.parseRunscript(scanner); err != nil {
			return err
		}
		goto again
	}
}
```
including the comment, but that AFAICS doesn’t handle EOF in `parseEnvironment`/`parseRunscript` correctly, calling `.Text()` after EOF. That’s not _invalid_ but the value is not obvious — I’d expect a trailing `%environment` line to cause an infinite loop, without checking in detail.

So, how about something vaguely like
```go
// enum state = parsingEnvironment/parsingRunscript/parsingOther
for scanner.Scan() {
	s := strings.TrimSpace(scanner.Text())
	switch {
	case s == "%environment": state = parsingEnvironment
	case s == "%runscript": state = parsingRunscript
	case strings.HasPrefix(s, "%"): state = parsingOther // Warn on unrecognized?
	case state == parsingEnvironment: image.parseEnvironmentEntry(s)
	case state == parsingRunscript: image.parseRunscriptEntry(s)
	default: // ignore s, or warn?
    }
}
```
?

Reply:
have you tested it?

Reply:
No. I’m not claiming that the code as is doesn’t work (well, apart from a trailing directive?), just that it looks very non-obvious.

If there are even more non-obvious gotchas that I didn’t notice by reading the code as is, those gotchas should be commented in detail and/or covered by unit tests.

Reply:
So I’m not sure which “it” I was supposed to test, but looking at https://github.com/mtrmac/image/tree/sif-wip :
- A trailing `%`… section header does not cause an infinite loop, because `scanner.Scan()`, after EOF, returns an empty string in `scanner.Text()`; AFAICS that’s totally undocumented (and possibly even contrary to the strict text of the existing documentation). So it’s good that there isn’t an infinite loop but we shouldn’t rely on that, unless there’s some guarantee I’m missing.
- A state-machine approach like the above does work, at least enough to pass a trivial smoke test I added.